ts/components/registration/stages/CreateAccount.tsx

```diff
@@ -0,0 +1,157 @@
+import { StringUtils } from '../../../session/utils';
+import { fromHex } from '../../../session/utils/String';
+import { trigger } from '../../../shims/events';
+import {
+  AccountCreation,
+  setAccountCreationStep,
+  setDisplayName,
+  setDisplayNameError,
+  setHexGeneratedPubKey,
+  setRecoveryPassword,
+} from '../../../state/onboarding/ducks/registration';
+import {
+  useDisplayName,
+  useDisplayNameError,
+  useRecoveryPassword,
+} from '../../../state/onboarding/selectors/registration';
+import {
+  generateMnemonic,
+  registerSingleDevice,
+  sessionGenerateKeyPair,
+} from '../../../util/accountManager';
+import { Storage, setSignWithRecoveryPhrase } from '../../../util/storage';
+import { Flex } from '../../basic/Flex';
+import { SessionButton, SessionButtonColor } from '../../basic/SessionButton';
+import { SpacerLG, SpacerSM } from '../../basic/Text';
+import { SessionInput } from '../../inputs';
+import { resetRegistration } from '../RegistrationStages';
+import { OnboardDescription, OnboardHeading } from '../components';
+import { BackButtonWithininContainer } from '../components/BackButton';
+import { displayNameIsValid, sanitizeDisplayNameOrToast } from '../utils';
+
+export type AccountDetails = {
+  recoveryPassword: string;
+  displayName?: string;
+};
+
+async function signUp(signUpDetails: AccountDetails) {
+  const { displayName, recoveryPassword } = signUpDetails;
+
+  try {
+    const validDisplayName = displayNameIsValid(displayName);
+    await resetRegistration();
+    await registerSingleDevice(recoveryPassword, 'english', validDisplayName);
+    await Storage.put(SettingsKey.hasSyncedInitialConfigurationItem, Date.now());
+    await setSignWithRecoveryPhrase(false);
+    trigger('openInbox');
+  } catch (e) {
+    await resetRegistration();
+    throw e;
+  }
+}
+
+export const CreateAccount = () => {
+  const recoveryPassword = useRecoveryPassword();
+  const displayName = useDisplayName();
+  const displayNameError = useDisplayNameError();
+
+  const dispatch = useDispatch();
+
+  const generateMnemonicAndKeyPair = async () => {
+    if (recoveryPassword === '') {
+      const mnemonic = await generateMnemonic();
+
+      let seedHex = mnDecode(mnemonic);
+      // handle shorter than 32 bytes seeds
+      const privKeyHexLength = 32 * 2;
+      if (seedHex.length !== privKeyHexLength) {
+        seedHex = seedHex.concat('0'.repeat(32));
+        seedHex = seedHex.substring(0, privKeyHexLength);
+      }
+      const seed = fromHex(seedHex);
+      const keyPair = await sessionGenerateKeyPair(seed);
+      const newHexPubKey = StringUtils.decode(keyPair.pubKey, 'hex');
+
+      dispatch(setRecoveryPassword(mnemonic));
+      dispatch(setHexGeneratedPubKey(newHexPubKey)); // our 'frontend' account ID
+    }
+  };
+
+  useMount(() => {
+    void generateMnemonicAndKeyPair();
+  });
+
+  const signUpWithDetails = async () => {
+    if (!(!!displayName && !displayNameError)) {
```

Comment:
you check for that same exact condition  on line 151. You can probably remove it from down there, but keep this one.

Also, this is equivalent to 
!displayName || displayNameError
right?

Reply:
What I meant is that the whole component returns early if 
`!(!!displayName && !displayNameError)` is true

So the disabled state is never reached below (as we return early when that condition is true!)
You need to keep the one returning early, but not the one on the disabled state

Reply:
There is no early return in the component `isEmpty(displayName) || !isEmpty(displayNameError)` happens in `signUpWithDetails` and the button disabled state so we need both. Resolved